### PR TITLE
Feature add whitelist parameter issue 2548

### LINKED ISSUE #2548: Add whitelist parameter to some contrib.tensorboard_logger & contrib.clearml_logger handlers
## 🚀 Feature

Following issue #2328 , we could add `whitelist` parameter to these classes in `contrib.tensorboard_logger` and `contrib.clearml_logger` to be able to log only desired weights or gradients.

1. GradsHistHandler
2. GradsScalarHandler
3. WeightsScalarHandler
4. WeightsHistHandler

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Fixes #2548

Description:
I made some refactors first and then added `whitelist` parameter to specified classes in the issue.

Check list:

- [x] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by sadra-barikbin** [comment]
How do I get rid of old commits in PR?

File: ignite/contrib/handlers/clearml_logger.py [hunk 1]
```diff
@@ -425,19 +423,14 @@ class WeightsScalarHandler(BaseWeightsScalarHandler):
             )
     """
 
-    def __init__(self, model: Module, reduction: Callable = torch.norm, tag: Optional[str] = None):
```
Comment (vfdev-5) [inline ignite/contrib/handlers/clearml_logger.py:428]: Docstring should be updated as `whitelist` arg is missing.

File: ignite/contrib/handlers/clearml_logger.py [hunk 2]
```diff
@@ -475,18 +468,13 @@ class WeightsHistHandler(BaseWeightsHistHandler):
             )
     """
 
-    def __init__(self, model: Module, tag: Optional[str] = None):
```
Comment (vfdev-5) [inline ignite/contrib/handlers/clearml_logger.py:478]: Same here for missing `whitelist` arg in docstring

File: ignite/contrib/handlers/clearml_logger.py [hunk 3]
```diff
@@ -528,18 +516,13 @@ class GradsScalarHandler(BaseWeightsScalarHandler):
             )
     """
 
-    def __init__(self, model: Module, reduction: Callable = torch.norm, tag: Optional[str] = None):
```
Comment (vfdev-5) [inline ignite/contrib/handlers/clearml_logger.py:531]: Same here for missing `whitelist` arg in docstring

File: ignite/contrib/handlers/clearml_logger.py [hunk 4]
```diff
@@ -577,18 +560,13 @@ class GradsHistHandler(BaseWeightsHistHandler):
             )
     """
 
-    def __init__(self, model: Module, tag: Optional[str] = None):
```
Comment (vfdev-5) [inline ignite/contrib/handlers/clearml_logger.py:580]: Same here for missing `whitelist` arg in docstring

File: ignite/contrib/handlers/clearml_logger.py [hunk 5]
```diff
@@ -577,18 +560,13 @@ class GradsHistHandler(BaseWeightsHistHandler):
...
-        super(GradsHistHandler, self).__init__(model, tag=tag)
-
     def __call__(self, engine: Engine, logger: ClearMLLogger, event_name: Union[str, Events]) -> None:
         if not isinstance(logger, ClearMLLogger):
             raise RuntimeError("Handler 'GradsHistHandler' works only with ClearMLLogger")
 
         global_step = engine.state.get_event_attrib_value(event_name)
         tag_prefix = f"{self.tag}/" if self.tag else ""
-        for name, p in self.model.named_parameters():
-            if p.grad is None:
```
Comment (vfdev-5) [inline ignite/contrib/handlers/clearml_logger.py:590]: Why do you remote this `if p.grad is None` ?

File: ignite/contrib/handlers/tensorboard_logger.py [hunk 6]
```diff
@@ -378,27 +376,22 @@ class WeightsScalarHandler(BaseWeightsScalarHandler):
             )
     """
 
-    def __init__(self, model: nn.Module, reduction: Callable = torch.norm, tag: Optional[str] = None):
```
Comment (vfdev-5) [inline ignite/contrib/handlers/tensorboard_logger.py:381]: Same here for missing `whitelist` arg in docstring

File: ignite/contrib/handlers/tensorboard_logger.py [hunk 7]
```diff
@@ -378,27 +376,22 @@ class WeightsScalarHandler(BaseWeightsScalarHandler):
...
-
     def __call__(self, engine: Engine, logger: TensorboardLogger, event_name: Union[str, Events]) -> None:
 
         if not isinstance(logger, TensorboardLogger):
             raise RuntimeError("Handler 'WeightsScalarHandler' works only with TensorboardLogger")
 
         global_step = engine.state.get_event_attrib_value(event_name)
         tag_prefix = f"{self.tag}/" if self.tag else ""
-        for name, p in self.model.named_parameters():
-            if p.grad is None:
```
Comment (vfdev-5) [inline ignite/contrib/handlers/tensorboard_logger.py:392]: Same question here for removed if clause

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @sadra-barikbin 
I left few comments and questions

**Comment by vfdev-5** [comment]
> How do I get rid of old commits in PR?

You have to create new branch from updated master

File: ignite/contrib/handlers/tensorboard_logger.py (hunk 7 above)
Comment (sadra-barikbin) [reply ignite/contrib/handlers/tensorboard_logger.py:392]: I restored it.

File: ignite/contrib/handlers/clearml_logger.py (hunk 5 above)
Comment (sadra-barikbin) [reply ignite/contrib/handlers/clearml_logger.py:590]: I restored it.

**Comment by sadra-barikbin** [comment]
I restored 

```python
if p.grad is None:
    continue
```
for all `Grads*Handler`s.

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 8]
```diff
@@ -474,37 +474,52 @@ def _test(tag=None):
...
+        ],
+        any_order=True,
+    )
     assert mock_logger.writer.add_scalar.call_count == 2
-    assert norm_mock.call_count == 2
 
 
 def test_grads_hist_handler_wrong_setup():
 
-    with pytest.raises(TypeError, match="Argument model should be of type torch.nn.Module"):
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:505]: Why do you remove this check ?

**Comment by vfdev-5** [comment]
@sadra-barikbin can you check if everything works as expected for TB and clearml using mnist or cifar10 example ?

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py (hunk 8 above)
Comment (sadra-barikbin) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py:505]: As `GradHistHandler` and its brothers all subclass `BaseWeightsHandler`, I added `test_base_weights_handler_wrong_setup` to `test_base_logger` and did this test there.

**Comment by sadra-barikbin** [comment]
image
image
image

In Tensorboard, everything related to weights seems to be correct, but gradients are constantly equal to zero because notebook trainer's step function does `zero_grad` before any logger call.

**Comment by sadra-barikbin** [comment]
`reduction` in `*ScalarHandler`s is considered to be just a `Callable`. Isn't better to enforce it to be `Callable[[Tensor], Tensor]` in order to be able to take advantage of GPU speed in reduction operation?

**Comment by vfdev-5** [comment]
> In Tensorboard, everything related to weights seems to be correct, but gradients are constantly equal to zero because notebook trainer's step function does zero_grad before any logger call.

yes, this happens if using `create_supervised_trainer`, there is a PR to fix that but it misses tests: https://github.com/pytorch/ignite/pull/2470 if we could make it land, it would be great.

**Comment by vfdev-5** [comment]
> `reduction` in `*ScalarHandler`s is considered to be just a `Callable`. Isn't better to enforce it to be `Callable[[Tensor], Tensor]` in order to be able to take advantage of GPU speed in reduction operation?

Just `Callable` is generic and has `Callable[[Tensor], Tensor]` as a subset, right ? So, we can benefit from using GPU. Am I missing your point ?

**Comment by sadra-barikbin** [comment]
> > `reduction` in `*ScalarHandler`s is considered to be just a `Callable`. Isn't better to enforce it to be `Callable[[Tensor], Tensor]` in order to be able to take advantage of GPU speed in reduction operation?
> 
> Just `Callable` is generic and has `Callable[[Tensor], Tensor]` as a subset, right ? So, we can benefit from using GPU. Am I missing your point ?

No. You're right. So I do that.

**Comment by sadra-barikbin** [comment]
image
image
image

ClearML `*ScalarHandler` seemingly works as expected but histogram ones not. Plots tab in project page is empty.

**Comment by sadra-barikbin** [comment]
image
image

TensorboardLogger works properly.

File: ignite/engine/__init__.py [hunk 9]
```diff
@@ -95,6 +95,8 @@ def supervised_training_step(
         )
 
     def update(engine: Engine, batch: Sequence[torch.Tensor]) -> Union[Any, Tuple[torch.Tensor]]:
+        if (engine.state.iteration - 1) % gradient_accumulation_steps == 0:
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:98]: @sadra-barikbin I prefer to finalize already open PR. Can you make another branch and use commits from https://github.com/pytorch/ignite/pull/2470 and add tests I asked in the PR. Such that we merge your PR first and update this one right after that.

File: tests/ignite/contrib/handlers/test_clearml_logger.py [hunk 10]
```diff
@@ -331,16 +331,7 @@ def global_step_transform(*args, **kwargs):
 
 def test_weights_scalar_handler_wrong_setup():
 
-    with pytest.raises(TypeError, match="Argument model should be of type torch.nn.Module"):
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_clearml_logger.py:334]: Why do you remove these checks ?

File: tests/ignite/contrib/handlers/test_clearml_logger.py [hunk 11]
```diff
@@ -381,45 +372,54 @@ def _test(tag=None):
...
+            call(title="model/weights_norm/fc2", value=ANY, series="bias", iteration=mock_engine.state.epoch),
+        ],
+        any_order=True,
+    )
     assert mock_logger.clearml_logger.report_scalar.call_count == 2
 
 
 def test_weights_hist_handler_wrong_setup():
 
-    with pytest.raises(TypeError, match="Argument model should be of type torch.nn.Module"):
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_clearml_logger.py:420]: Same question here: why ?

File: tests/ignite/contrib/handlers/test_clearml_logger.py (hunk 10 above)
Comment (vfdev-5) [reply tests/ignite/contrib/handlers/test_clearml_logger.py:334]: OK, I got your answer from below comment -> test_base_weights_handler_wrong_setup

**Comment by vfdev-5** [comment]
@sadra-barikbin can you please resolve the conflict and update this PR

File: tests/ignite/engine/test_create_supervised.py [hunk 12]
```diff
@@ -125,6 +125,37 @@ def _():
                 trainer.run(data)
 
 
+def _test_create_supervised_trainer_have_grad_after_iteration(
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:128]: This should be removed in this PR

File: ignite/contrib/handlers/tensorboard_logger.py [hunk 13]
```diff
@@ -522,39 +538,78 @@ class GradsScalarHandler(BaseWeightsScalarHandler):
...
         tag_prefix = f"{self.tag}/" if self.tag else ""
-        for name, p in self.model.named_parameters():
+        for name, p in self.weights:
             if p.grad is None:
                 continue
 
             name = name.replace(".", "/")
             logger.writer.add_scalar(
-                f"{tag_prefix}grads_{self.reduction.__name__}/{name}", self.reduction(p.grad), global_step
+                f"{tag_prefix}grads_{self.reduction.__name__}/{name}", self.reduction(p.grad).item(), global_step
```
Comment (vfdev-5) [inline ignite/contrib/handlers/tensorboard_logger.py:597]: `self.reduction(p.grad)` is not always Tensor and can be also float. Thus calling `.item()` will fail. I think it is safe to remove `.item` here.

File: ignite/contrib/handlers/base_logger.py [hunk 14]
```diff
@@ -136,14 +173,20 @@ def key_str_tf(tag: str, name: str, *args: str) -> str:
...
+    Helper handler to log model's weights or gradients as scalars.
     """
 
-    def __init__(self, model: nn.Module, reduction: Callable = torch.norm, tag: Optional[str] = None):
-        if not isinstance(model, torch.nn.Module):
-            raise TypeError(f"Argument model should be of type torch.nn.Module, but given {type(model)}")
+    def __init__(
+        self,
+        model: nn.Module,
+        reduction: Callable[[torch.Tensor], torch.Tensor] = torch.norm,
```
Comment (vfdev-5) [inline ignite/contrib/handlers/base_logger.py:184]: `Callable[[torch.Tensor], torch.Tensor]` is incorrect type hint as it can be 0D tensor or a number, see line 199 (below):
```python
if not (isinstance(o, numbers.Number) or _is_0D_tensor(o)):
```

File: ignite/contrib/handlers/clearml_logger.py [hunk 15]
```diff
@@ -435,25 +478,29 @@ def __call__(self, engine: Engine, logger: ClearMLLogger, event_name: Union[str,
...
-                continue
+        for name, p in self.weights:
 
             title_name, _, series_name = name.partition(".")
             logger.clearml_logger.report_scalar(
                 title=f"{tag_prefix}weights_{self.reduction.__name__}/{title_name}",
                 series=series_name,
-                value=self.reduction(p.data),
+                value=self.reduction(p.data.detach()).cpu(),
```
Comment (vfdev-5) [inline ignite/contrib/handlers/clearml_logger.py]: Here you made this assumption on the output as tensor. This code will fail if output is python float and which is a possible reduction function output. I do not think that we want to change that.

Comment (sadra-barikbin) [reply ignite/contrib/handlers/clearml_logger.py]: I changed `reduction` type hint to `Callable[[torch.Tensor], Any]`.

Comment (vfdev-5) [reply ignite/contrib/handlers/clearml_logger.py]: How about replacing `Any` by `Union[Number, torch.Tensor]` ?

Comment (sadra-barikbin) [reply ignite/contrib/handlers/clearml_logger.py]: Yes. That sounds better.

Comment (vfdev-5) [reply ignite/contrib/handlers/clearml_logger.py]: We may fix mypy issue with visdom as : 
1) ignoring type mismatch
2) Use `Union[float, torch.Tensor]`
I let you decide

File: ignite/contrib/handlers/base_logger.py [hunk 16]
```diff
@@ -191,7 +191,7 @@ def __init__(
         if not callable(reduction):
             raise TypeError(f"Argument reduction should be callable, but given {type(reduction)}")
 
-        def _is_0D_tensor(t: torch.Tensor) -> bool:
+        def _is_0D_tensor(t: Any) -> bool:
```
Comment (vfdev-5) [inline ignite/contrib/handlers/base_logger.py:194]: Why this change ?

File: ignite/contrib/handlers/clearml_logger.py (hunk 15 above)
Comment (sadra-barikbin) [reply ignite/contrib/handlers/clearml_logger.py]: The safer, the better.
I've got to change `_is_0D_tensor` parameter type hint as well. It wants to check if its argument is a tensor and beside that a 1D tensor so being `Any` is a reasonable assumption.

File: ignite/contrib/handlers/base_logger.py (hunk 16 above)
Comment (sadra-barikbin) [reply ignite/contrib/handlers/base_logger.py:194]: I've got to change _is_0D_tensor parameter type hint as well. It wants to check if its argument is a tensor and beside that a 1D tensor so being Any is a reasonable assumption.

Comment (sadra-barikbin) [reply ignite/contrib/handlers/base_logger.py:194]: Otherwise Mypy raised error.
```bash
error: Argument 1 to "_is_0D_tensor" has incompatible type
"Union[float, Tensor]"; expected "Tensor"  [arg-type]
```

Comment (vfdev-5) [reply ignite/contrib/handlers/base_logger.py:194]: OK, it is actually correct that previous type hint is wrong. Thanks

**Review by vfdev-5 — APPROVED** [review]
@sadra-barikbin Thanks, lgtm

**Comment by vfdev-5** [comment]
> ClearML *ScalarHandler seemingly works as expected but histogram ones not. Plots tab in project page is empty.

@sadra-barikbin you were reporting that previously. Can you check again if this is still a problem with the latest clearml and see if it related to our code or not. cc @bmartinn

**Comment by sadra-barikbin** [comment]
ClearML histograms were still missing. There was an issue in calling `WeightsGradientHistHelper` on logger construction. This object is used in logging histograms. Its `report_freq` parameter's default value is 100 so among every 100 calls 99 of them are returned without any actual log. I changed it to 1 and histograms are now plotted correctly.

[Bug place](https://github.com/pytorch/ignite/blob/a74b5b06335486d5d5e3178fe1912f48a85408b1/ignite/contrib/handlers/clearml_logger.py#L118)
Final result:
image

The `Iteration 0` that is in plot title is due to internal stuff of `WeightsGradientHistHelper`. As we call `add_histogram`, it does some processes and finally calls clearml logger's `add_surface` method with our arguments `title` and `series` unchanged and also with another argument, `iteration` hard-coded to zero.

**Comment by vfdev-5** [comment]
@sadra-barikbin thanks a lot for investigations ! It is very helpful !

> As we call add_histogram, it does some processes and finally calls clearml logger's add_surface method with our arguments title and series unchanged and also with another argument, iteration hard-coded to zero.

So, we call add_histogram:
https://github.com/pytorch/ignite/blob/a74b5b06335486d5d5e3178fe1912f48a85408b1/ignite/contrib/handlers/clearml_logger.py#L493-L498
and define `step` as `global_step` but step is not propagated to `add_surface` ?
